### ISSUE TITLE
How to choose the radius size when I need to look for neighbours of each point 

### ISSUE THREAD

**Issue by Luan-Zhaoliang** [HuguesTHOMAS/KPConv-PyTorch#5]
Hi,
I'm replicating your code, but using a different data set. I want to know how I can set the search radius to find neighbours points. In your paper, you just said: **_where σ is the inﬂuence distance of the kernel points, and will be chosen according to the input density_**, whereas in your code **_(config.py)_** you set some default values for the radius and density parameters, I wanna know if there was any reference when I'm going to set the search radius?

**Comment by HuguesTHOMAS** [comment]
Hi @Luan-Zhaoliang,

the parameters you are interested in are defined here:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/ccb820bbb7a18e4642c317bc8a80d26164f7f3c1/train_S3DIS.py#L102-L112

 * `first_subsampling_dl` controls the size of the subsampling grid cells in meters. So here it is set to 3cm.
 * `conv_radius` is the radius of the neighborhood (for rigid KPConv). The unit is the number of grid cells. So here it is set to 2.5 grid cells = 7.5cm.
* `deform_radius` is the same as conv radius but for deformable KPConv (ignore it if you do not use deformable layers)
* `KP_extent` is the the inﬂuence distance of the kernel points. This is the σ in our paper, but it does not affect the effective radius of neighbors. The unit is also the number of grid cells. Here I used 1.2 which is a bit larger that the value 1.0 used in the paper.

I hope this helps.

**Comment by Luan-Zhaoliang** [comment]
Hi Thomas,

Very thanks for your reply. It is very helpful for me.

regards, Zhaoliang


________________________________
发件人: Hugues THOMAS <notifications@github.com>
发送时间: 2020年5月11日 16:33
收件人: HuguesTHOMAS/KPConv-PyTorch <KPConv-PyTorch@noreply.github.com>
抄送: Zhaoliang Luan <z.luan@qmul.ac.uk>; Mention <mention@noreply.github.com>
主题: Re: [HuguesTHOMAS/KPConv-PyTorch] How to choose the radius size when I need to look for neighbours of each point (#5)


Hi @Luan-Zhaoliang<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FLuan-Zhaoliang&data=02%7C01%7C%7Ca1df63edaca64c1d3c7108d7f5c09956%7C569df091b01340e386eebd9cb9e25814%7C0%7C1%7C637248079864448982&sdata=ZxoPcrBVirSomSYDX%2BSE2YnU95qLzqNjqKqJhvXee7U%3D&reserved=0>,

the parameters you are interested in are defined here:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/ccb820bbb7a18e4642c317bc8a80d26164f7f3c1/train_S3DIS.py#L102-L112<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHuguesTHOMAS%2FKPConv-PyTorch%2Fblob%2Fccb820bbb7a18e4642c317bc8a80d26164f7f3c1%2Ftrain_S3DIS.py%23L102-L112&data=02%7C01%7C%7Ca1df63edaca64c1d3c7108d7f5c09956%7C569df091b01340e386eebd9cb9e25814%7C0%7C1%7C637248079864448982&sdata=SXeE79yEHabFpIgCM92FNXIQcgdgAE%2Fc1eI83%2FXnmgQ%3D&reserved=0>

  *   first_subsampling_dl controls the size of the subsampling grid cells in meters. So here it is set to 3cm.
  *   conv_radius is the radius of the neighborhood (for rigid KPConv). The unit is the number of grid cells. So here it is set to 2.5 grid cells = 7.5cm.
  *   deform_radius is the same as conv radius but for deformable KPConv (ignore it if you do not use deformable layers)
  *   KP_extent is the the inﬂuence distance of the kernel points. This is the σ in our paper, but it does not affect the effective radius of neighbors. The unit is also the number of grid cells. Here I used 1.2 which is a bit larger that the value 1.0 used in the paper.

I hope this helps.

**Comment by jedolb** [comment]
Hi, thank you for sharing this great work ! I have a question related to what you said in this issue. If I have a denser dataset, in which direction should I change these parameters? Should I increase or decrease their values?

**Comment by HuguesTHOMAS** [comment]
Hi @jedolb,

It depends on what kind of objects you are trying to detect. Having a denser dataset allows you to reduce the `first_subsampling_dl`, and therefore have shapes with finer details. To avoid having your memory explode, you also want to reduce the `in_radius` (which was not listed above and controls the size of the input sphere) parameter accordingly. In that case your network will be better on small objects but could lose performances on very big objects.

If your dataset is denser but still have big objects, then maybe this density is not useful to you and you can keep a higher `first_subsampling_dl` and `in_radius`.

In any case, the other parameters should remain the same as a change of dataset density will not affect them.

**Comment by jedolb** [comment]
Thank you very much for your quick answer and your explanations !

Actually, I'm trying to train a KPConv network on accumulated Lidar data. I accumulated the sequences of the SemanticKitti dataset, and I've cut out sections of 20m2. So I feed the network with these sections. One section looks like this : 
![acc4_11](https://user-images.githubusercontent.com/67365055/89200506-685b5d80-d5b0-11ea-8528-216f61e98b48.png)

I made a first training with the parameters of train_SemanticKitti.py . The results are not that bad, the validation accuracy is 0,8.  Here a prediction : 
![acc4_11_pred](https://user-images.githubusercontent.com/67365055/89200589-89bc4980-d5b0-11ea-8dac-12994e4c70ec.png)

Some classes are well detected like cars or buildings. But he can't tell the difference between road, sidewalk and terrain. I guess I can increase performance. I've big objects, but I may have enough detail to reduce first_subsampling_dl and in_radius. There's no harm in trying :)

**Comment by HuguesTHOMAS** [comment]
Exactly, for this kind of application where you are not sure which values are the best, you can just try a bunch of them.

I would advise to start from the parameters of NPM3D dataset which is a similar situation. and then experiment to finetune the values. In any case, road, sidewalk and terrain are very challenging classes when you don't have colors, as they have nearly the same geometric shape. You could try to use the lidar intensity value but I think this is not going to help.

Good luck with your problem.

**Comment by jedolb** [comment]
I will first try with the parameters of NPM3D then 

Thanks again for your help! It is very helpful for me
